torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,220 @@
+
+      inp_avg[index_in_chunk] = self_avg;
+      inp_var[index_in_chunk] = self_var;
+    }
+    inp_N = updated_N;
+  }
+
+  // At this point, each thread has partial final results in
+  // inp[0:chunk_size]
+
+  // Upload to smem
+#pragma unroll
+  for (int i = 0; i < chunk_size; ++i) {
+    // int smem_offset = tid + i * BDIMX * blockDim.y;
+    int smem_offset = 0;
+    smem_offset += wid * 32;
+    smem_offset += ((warp_tidy + wid) % wdimy) * BDIMX;
+    smem_offset += threadIdx.x;
+    smem_offset += i * BDIMX * BDIMY;
+    smem[smem_offset] = inp_avg[i];
+    smem[smem_var_offset + smem_offset] = inp_var[i];
+    // Upload N only when threadIdx.x == 0 && chunk_index == 0
+    if (threadIdx.x == 0 && i == 0 && warp_tidy == 0) {
+      reinterpret_cast<nvfuser_index_t*>(smem + smem_N_offset)[wid] = inp_N;
+    }
+  }
+
+  __syncthreads();
+
+  // Final reduction. Assign one warp per group. The vector width is
+  // likely at least 4, so at least 4 warps should be used, which is
+  // enough to occupy an SM. When NumVals=8, it might be more
+  // efficient to use just 4 warps with each warp taking care of two
+  // groups, but the difference would be pretty small.
+
+  // Also, the number of warps should be at least 8 and can be 16
+  // too. NumVals should be 8 at largest, so it's always num_warps >=
+  // NumVals.
+  static_assert(
+      num_warps >= NumVals,
+      "Number of warps must be at least as large as NumVals");
+
+  DataType avg = 0;
+  DataType var = 0;
+  nvfuser_index_t N = 0;
+
+  // First, per-thread reduction
+  if (wid < NumVals) {
+#pragma unroll
+    for (int i = warp_tidy; i < num_warps; i += wdimy) {
+      int offset = 0;
+      // x=tidx
+      offset += threadIdx.x;
+      // y=i
+      offset += i * 32;
+      // group=wid
+      // offset += (wid / chunk_size) * BDIMX;
+      int offset_y = wid / chunk_size;
+      // swizzle
+      offset_y = (offset_y + i) % wdimy;
+      offset += offset_y * BDIMX;
+
+      offset += (wid % chunk_size) * BDIMX * BDIMY;
+
+      DataType avg_smem = smem[offset];
+      DataType var_smem = smem[smem_var_offset + offset];
+      nvfuser_index_t N_smem =
+          reinterpret_cast<nvfuser_index_t*>(&smem[smem_N_offset])[i];
+
+      welfordCombine(avg, var, N, avg_smem, var_smem, N_smem);
+    }
+  }
+
+  __syncthreads();
+
+  if (wid >= NumVals) {
+    WelfordTriplet<DataType> out = {0, 0, 0};
+    return out;
+  }
+
+  // Reduction within each warp
+#pragma unroll
+  for (int lane_mask = 16; lane_mask >= BDIMX; lane_mask /= 2) {
+    auto avg_peer = __shfl_xor_sync(0xffffffff, avg, lane_mask);
+    auto var_peer = __shfl_xor_sync(0xffffffff, var, lane_mask);
+    auto N_peer = __shfl_xor_sync(0xffffffff, N, lane_mask);
+
+    welfordCombine(avg, var, N, avg_peer, var_peer, N_peer);
+  }
+
+  // Each warp only has a valid result of group_id == wid
+  WelfordTriplet<DataType> out = {avg, var, N};
+  return out;
+}
```

Comment:
Having trouble following the remappings in the reduction so having trouble checking over the indexing math. I don't think I'm going to go further into it, but I hope it's all correct...

Reply:
Thanks for checking. Added more comments.